src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -1931,29 +1931,32 @@ protected void EnsureNullabilityAnalysisPerformedIfNecessary()
             var bindableRoot = GetBindableSyntaxNode(Root);
             using var upgradeableLock = _nodeMapLock.DisposableUpgradeableRead();
 
-            if (_guardedNodeMap.ContainsKey(bindableRoot)
+            // If there are already nodes in the map, then we've already done work here. Since
+            // EnsureNullabilityAnalysis is guaranteed to have run first, that means we've
+            // already bound the root node and we can just exit. We can't just assert that Root
+            // is in the map, as there are some models for which there is no BoundNode for the
+            // Root elements (such as fields, where the root is a VariableDeclarator but the
+            // first BoundNode corresponds to the underlying EqualsValueSyntax of the initializer)
+            if (_guardedNodeMap.Count > 0)
```

Comment:
>if (_guardedNodeMap.Count > 0) [](start = 12, length = 30)

Could this be a race? We might have started populating the map, but haven't finished. The request might start binding on its own because of the fact that the map is incomplete. I think we might want to introduce a new API to get the "effective bindable root" node. Each specific member model has to implement it. For example, InintializerSemanticModel.Bind performs adjustments like that. Perhaps it could be implemented in terms of the new API.

Reply:
> Could this be a race? We might have started populating the map, but haven't finished.

This can't be observed in a partially-added stated because the next thing we do here is enter a write lock, which does not allow other reads or upgradable reads. The specific pattern we were seeing with field initializers was this:

1. Threads 1 and 2 enter this method, and both check `_lazySnapshotManager` for `null`. Both encounter null and move to acquiring an upgradeable reader lock.
2. ReaderWriterLockSlim only allows 1 upgradeable lock at a time, so thread 1 gets the lock and 2 waits.
3. Thread 1 binds the body and adds all found bound nodes. This does not include a `BoundNode` for the `VariableDeclaratorSyntax` node because it doesn't exist.
4. Thread 1 releases the write lock.
5. Thread 2 wakes up and acquires the upgradable read. It observes that there are now nodes in the map (from Thread 1), but there is no `BoundNode` for the `Root` node.

>  I think we might want to introduce a new API to get the "effective bindable root" node. Each specific member model has to implement it. For example, InintializerSemanticModel.Bind performs adjustments like that.

We certainly could implement an API like that. However, beyond simplifying this assert I'm not sure that it would simplify the code here.

---
In reply to: [454648399](https://github.com/dotnet/roslyn/pull/45967#discussion_r454648399) [](ancestors = 454648399)